src/udp/socket.rs

```diff
@@ -1158,4 +1162,133 @@ mod test {
+                            drop(server_drop_tx_0);
+                        })
+                    }));
+                    unwrap!(res)
+                }),
+            );
+            let node_1 = node::nat_v4(
+                NatV4Builder::default(),
+                node::endpoint_v4(move |_ip| {
+                    let mut core = unwrap!(Core::new());
+                    let handle = core.handle();
+
+                    let res = core.run(future::lazy(|| {
+                        Timeout::new(start_delay, &handle)
+                        .infallible()
+                        .and_then(|()| server_addr_rx_1)
+                        .map_err(|e| panic!("error getting server addr: {}", e))
+                        .map(|server_addr| {
+                            let p2p = P2p::default();
+                            p2p.add_udp_traversal_server(&server_addr);
+                            p2p
+                        })
+                        .and_then(|p2p| {
+                            UdpSocket::rendezvous_connect(ch1, &handle, &p2p)
+                            .map_err(|e| panic!("rendezvous connect error: {}", e))
+                        })
+                        .map(|_socket| {
+                            trace!("connected peer 1");
+                            drop(server_drop_tx_1);
+                        })
+                    }));
+                    unwrap!(res)
+                }),
+            );
+
+            let server_node = node::latency_v4(
+                Duration::from_millis(100),
+                Duration::from_millis(10),
+                node::hops_v4(2, server_node),
+            );
+            let node_0 = node::latency_v4(
+                Duration::from_millis(200),
+                Duration::from_millis(20),
+                node::hops_v4(3, node_0),
+            );
+            let node_0 = node::latency_v4(
+                Duration::from_millis(200),
+                Duration::from_millis(20),
+                node::hops_v4(3, node_0),
+            );
+
+            let network = node::router_v4((
+                server_node,
+                node_0,
+                node_1,
+            ));
+
+            let (join_handle, _plug) = netsim::spawn::network_v4(&handle, SubnetV4::global(), network);
+
+            future_utils::thread_future(|| unwrap!(join_handle.join()))
+            .map(|((), (), ())| ())
+        }));
+        res.void_unwrap()
+    }
+
+    #[test]
+    fn udp_rendezvous_connect_between_natted_hosts_with_no_delay() {
+        udp_rendezvous_connect_between_natted_hosts(Duration::from_secs(0));
+    }
+
+    #[test]
+    fn udp_rendezvous_connect_between_natted_hosts_with_short_delay() {
+        udp_rendezvous_connect_between_natted_hosts(Duration::from_secs(5));
+    }
+
+    #[test]
+    fn udp_rendezvous_connect_between_natted_hosts_with_long_delay() {
+        udp_rendezvous_connect_between_natted_hosts(Duration::from_secs(60));
+    }
```

Comment:
uuu, this file is huge :d 
I was planning to move hole punching related code to `src/udp/hole_punching.rs` - just a note for future reference :)

Reply:
Yeah, it would be good to slit it up somehow..